### DISCUSSION TITLE
macOS GPU resource sharing

### DISCUSSION THREAD

**Discussion by mzenz** [containers/podman#22580] [Q&A]
Is it possible to share GPU resources with Podman containers, similar to how Nvidia GPU resources can be shared with containers on Linux/Windows?

Please let us know, I haven't been able to find any info on how to do so.

**Comment by rhatdan** [comment]
We are working on this and it should be available in podman 5.1.

https://github.com/containers/podman/pull/22491

**Reply by DrogeAppel** [reply]
Is this future available right now because 5.1 is out now?

**Reply by rhatdan** [reply]
Well  yes and know. You can switch to krunkit, but need to get the bugfix in podman 5.1.1

**Reply by rhatdan** [reply]
https://github.com/containers/podman/releases/tag/v5.1.1

**Comment by DrogeAppel** [comment]
So install version 5.1.1 and then with Krunkit I can have a container use GPU resources.

Do you have some tutorial for this?

**Reply by rhatdan** [reply]
Not that I know of.
@slp @baude @mheon Any ideas?

**Reply by slp** [reply]
I wrote a [blog post](https://sinrega.org/2024-03-06-enabling-containers-gpu-macos/) about it, but it was from before support was merged into podman. Nowadays is much simpler:

1. Configure `libkrun` as provider. You can do this temporarily by exporting a environment variable (`export CONTAINERS_MACHINE_PROVIDER="libkrun"`) or permanently by adding this to `~/.config/containers/containers.conf`:
```
[machine]
  provider = "libkrun"
```

2. Create a VM for this provider the usual way (`podman machine init`).

Now you'll have on your podman VM some `/dev/dri/*` entries corresponding to the virtio-gpu device that provides GPU acceleration in the guest.

To make use of them, you still need a patched version of Mesa. We're working on getting this upstream (it's just two small fixes) to make it work with the Mesa version shipped with the distros. Meanwhile, please refer to the [Running an AI workload offloaded to the GPU](https://sinrega.org/2024-03-06-enabling-containers-gpu-macos/#running-an-ai-workload-offloaded-to-the-gpu) section of the blog post.

**Reply by tsailiming** [reply]
I'm on podman 5.1.1. When using `libkrun` as a provider, do we need to ` brew tap slp/krun`?

I had to add `krunkit` otherwise podman will complain
```
$ podman machine start
Starting machine "podman-machine-default"
ERRO[0000] process 29635 has not ended
Error: exec: "krunkit": executable file not found in $PATH
```

This is what I have installed: 
```
$ brew list --full-name | grep -i krun
slp/krunkit/krunkit
slp/krunkit/libkrun-efi
slp/krunkit/virglrenderer
```

**Reply by slp** [reply]
Yes, it's in the process of getting bundled: https://github.com/containers/podman/pull/23079

**Reply by sroecker** [reply]
I guess this can be marked as answered now since libkrun is bundled and the process is documented https://podman-desktop.io/docs/podman/gpu

**Reply by zoidyzoidzoid** [reply]
Hey folks! I recently installed the latest versions of podman and podman-desktop via brew and still needed to `brew install slp/krunkit/krunkit`.

**Reply by gclawes** [reply]
I'm seeing the same as @zoidyzoidzoid, I don't see `libkrun` anywhere in the podman-desktop install for 1.16.0 (installed from HomeBrew on an M1):
```
MacBook-Pro :: ~ % find /usr/local /opt/homebrew /Applications/Podman\ Desktop.app | grep -i krun
MacBook-Pro :: ~ %
```

**Reply by zoidyzoidzoid** [reply]
I think I found the issue. The brew Formula [here](https://github.com/Homebrew/homebrew-core/blob/322b367312ca276dd30ea95404488bb983d8c46f/Formula/p/podman.rb) needs to be updated to include the krun binary, and maybe also the other files introduced in https://github.com/containers/podman/pull/23079 

They're in the .pkg 
![image](https://github.com/user-attachments/assets/354a53b1-129c-499c-91c0-3df88fcaa8ec)
![image](https://github.com/user-attachments/assets/005d0693-b984-42a5-89d2-fd0d2687409b)

**Reply by slp** [reply]
@zoidyzoidzoid You're right. The podman releases in GitHub do bundle krunkit, but the brew version doesn't. If you only install podman-desktop but not podman, the first will detect that it's missing and will download it from a release, which does have krunkit bundled. If you install both from brew, podman-desktop detects that's already available and won't offer downloading it.

As you said, I think the brew formula for podman should be updated to include krunkit (cc/ @chenrui333).
